python/ray/train/mosaic/mosaic_trainer.py

```diff
@@ -0,0 +1,168 @@
+    ...     train_dataloader = torch.utils.data.DataLoader(
+    ...         config.pop("train_dataset"),
+    ...         batch_size=batch_size_per_worker
+    ...     )
+    ...     train_dataloader = ray.train.torch.prepare_data_loader(train_dataloader)
+    ...
+    ...     # prepare optimizer
+    ...     optimizer = composer.optim.DecoupledSGDW(
+    ...         model.parameters(),
+    ...         lr=0.05,
+    ...         momentum=0.9,
+    ...         weight_decay=2.0e-3,
+    ...     )
+    ...
+    ...     return composer.trainer.Trainer(
+    ...         model=model,
+    ...         train_dataloader=train_dataloader,
+    ...         optimizers=optimizer,
+    ...         **config
+    ...     )
+    ...
+    >>> scaling_config = ScalingConfig(num_workers=2, use_gpu=True)
+    >>> trainer_init_config = {
+    ...     "max_duration": "1ba",
+    ...     "train_dataset": train_dataset,
+    ...     "loggers": [InMemoryLogger()],
+    ...     "algorithms": [LabelSmoothing()],
+    ... }
+    ...
+    >>> trainer = MosaicTrainer(
+    ...     trainer_init_per_worker=trainer_init_per_worker,
+    ...     trainer_init_config=trainer_init_config,
+    ...     scaling_config=scaling_config,
+    ... )
+    ...
+    >>> trainer.fit()
+
+
+    Args:
+        trainer_init_per_worker: The function that returns an instantiated
+            ``composer.Trainer`` object and takes in an optional configuration
+            dictionary as an argument.
+        datasets: Any Ray Datasets to use for training.
+        trainer_init_config: Configurations to pass into ``trainer_init_per_worker`` as
+            kwargs.
+        torch_config: Configuration for setting up the PyTorch backend. If set to
+            None, use the default configuration. This replaces the ``backend_config``
+            arg of ``DataParallelTrainer``. Same as in ``TorchTrainer``.
+        scaling_config: Configuration for how to scale data parallel training.
+        dataset_config: Configuration for dataset ingest.
+        run_config: Configuration for the execution of the training run.
+        preprocessor: A ray.data.Preprocessor to preprocess the
+            provided datasets.
+        resume_from_checkpoint: A MosiacCheckpoint to resume training from.
+    """
+
+    def __init__(
+        self,
+        trainer_init_per_worker: Callable[[Optional[Dict]], Trainer],
+        *,
+        datasets: Optional[Dict[str, GenDataset]] = None,
+        trainer_init_config: Optional[Dict] = None,
+        torch_config: Optional[TorchConfig] = None,
+        scaling_config: Optional[ScalingConfig] = None,
+        dataset_config: Optional[Dict[str, DatasetConfig]] = None,
+        run_config: Optional[RunConfig] = None,
+        preprocessor: Optional["Preprocessor"] = None,
+        resume_from_checkpoint: Optional[Checkpoint] = None,
+    ):
+
+        self._validate_trainer_init_per_worker(
+            trainer_init_per_worker, "trainer_init_per_worker"
+        )
+
```

Comment:
If none of the Mosaic Loggers work with AIR, I'm wondering if we should raise an error if loggers are defined in `trainer_init_config` and redirect users to the AIR Callbacks instead.

Reply:
ditto https://github.com/ray-project/ray/pull/29237#discussion_r995115555

Reply:
Yeah the code may not error, but the behavior still may not be what users expect. Unless we can support the full functionality of loggers, then I think it's best to be explicit here and not allow users to pass in Mosaic Loggers, but rather redirect them to AIR Callbacks which provide equivalent functionality.

Reply:
@xwjiang2010 would like to hear your thoughts on this? Do you think it makes sense to explicitly steer users towards the AIR equivalents?

@Yard1 how did we handle this for HuggingfaceTrainer?

Reply:
@amogkam for HuggingfaceTrainer we do
```
    # ensure no HF logging callbacks are added
    # aside from doubling functionality with our callbacks,
    # the Wandb callbacks causes training to freeze
    integration_callbacks = transformers.trainer.get_reporting_integration_callbacks(
        trainer.args.report_to
    )
    for callback in integration_callbacks:
        trainer.pop_callback(callback)

    trainer.add_callback(TrainReportCallback)
```

I think we should do the same with Mosaic. Providing basic functionality is essential.

Reply:
Yeah, we should just use our callbacks and prevent the usage of Mosaic callbacks.

Reply:
We are now removing all composer loggers (but leaving other callbacks, since callbacks are used to provide data to be logged whereas loggers handle actual reporting, which can be replaced with Ray loggers).